server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java

```diff
@@ -129,101 +109,106 @@ public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Vo
                 "Timestamp to be pinned is less than current timestamp - value of cluster.remote_store.pinned_timestamps.lookback_interval"
             );
         }
-        updatePinning(pinnedTimestamps -> pinnedTimestamps.pin(timestamp, pinningEntity), listener);
+        long startTime = System.nanoTime();
+        try {
+            logger.debug("Pinning timestamp = {} against entity = {}", timestamp, pinningEntity);
+            blobContainer.writeBlob(getBlobName(timestamp, pinningEntity), new ByteArrayInputStream(new byte[0]), 0, true);
+            long elapsedTime = System.nanoTime() - startTime;
+            if (elapsedTime > RemoteStoreSettings.getPinnedTimestampsLookbackInterval().nanos()) {
+                String errorMessage = String.format(
+                    Locale.ROOT,
+                    "Timestamp pinning took %s nanoseconds which is more than limit of %s nanoseconds, failing the operation",
+                    elapsedTime,
+                    RemoteStoreSettings.getPinnedTimestampsLookbackInterval().nanos()
+                );
+                unpinTimestamp(timestamp, pinningEntity, ActionListener.wrap(() -> listener.onFailure(new RuntimeException(errorMessage))));
```

Comment:
- can it happen that pin passes and unpin fails? do we need to plan about some background cleanup thread in future?
- should the unpinning we done on an async executor thread, so as to not block the caller for long? if pinning took longer i guess there is a high change unpinning also will take longer

Reply:
>can it happen that pin passes and unpin fails? do we need to plan about some background cleanup thread in future?